### ISSUE TITLE
Ad hides data

### ISSUE THREAD

**Issue by JonnyCiel** [googleads/googleads-mobile-android-examples#86]
Hello. 

I am using Native Ad Express in my Recyclerview. However, If i have 10 items and i show the ads, the place where the ad is shown replaces the item from the data set.

How can i avoid this issue?

Thank you in advance.

**Comment by stowy** [comment]
Hi @JonnyCiel thanks for your question :+1: 

**TL;DR** You should:
1. Check out [this tutorial on YouTube](https://www.youtube.com/watch?v=LZCZSeFTvyk&index=11&list=PLOU2XLYxmsIKX0pUJV3uqp6N3NeHwHh0c).
2. Update your recycler view adapter to include a `NativeExpressAdView` view type.
3. Insert the ads into your recycler view items array, so that the number of items is increased by the number of ads that you add into the list.
4. Ensure your adapter returns the right item count and view type for each piece of data (eg your apps data or an ad). 

For example, for step 3, see this code snippet from  [this example project (NativeExpressRecyclerViewExample)](https://github.com/googleads/googleads-mobile-android-examples/blob/master/advanced/NativeExpressRecyclerViewExample/app/src/main/java/com/google/android/gms/example/nativeexpressrecyclerviewexample/MainActivity.java#L80):

```java
    /**
     * Adds Native Express ads to the items list.
     */
    private void addNativeExpressAds() {

        // Loop through the items array and place a new Native Express ad in every ith position in
        // the items List.
        for (int i = 0; i <= mRecyclerViewItems.size(); i += ITEMS_PER_AD) {
            final NativeExpressAdView adView = new NativeExpressAdView(MainActivity.this);
            mRecyclerViewItems.add(i, adView);
        }
    }
```
Note, in the above code, the `adView` is inserted into the `mRecyclerViewItems` array, which increases the size of the items array to include the ads. So the ads are part of the effective data source for the RecyclerView. 

For step 4, here's a snippet from the [RecyclerViewAdapter](https://github.com/googleads/googleads-mobile-android-examples/blob/master/advanced/NativeExpressRecyclerViewExample/app/src/main/java/com/google/android/gms/example/nativeexpressrecyclerviewexample/RecyclerViewAdapter.java#L73):

```
    @Override
    public int getItemCount() {
        return mRecyclerViewItems.size();
    }

    /**
     * Determines the view type for the given position.
     */
    @Override
    public int getItemViewType(int position) {
        return (position % MainActivity.ITEMS_PER_AD == 0) ? NATIVE_EXPRESS_AD_VIEW_TYPE
                : MENU_ITEM_VIEW_TYPE;
    }
```

In the above `getItemViewType` method, we know that every `ITEMS_PER_AD` number in the `mRecyclerViewItems` array will be an ad, and we return the `NATIVE_EXPRESS_AD_VIEW_TYPE` for that position. Otherwise we return our `MENU_ITEM_VIEW_TYPE` which is the data type for our apps content.

From the above example project look in particular at the [RecyclerViewAdapter](https://github.com/googleads/googleads-mobile-android-examples/blob/master/advanced/NativeExpressRecyclerViewExample/app/src/main/java/com/google/android/gms/example/nativeexpressrecyclerviewexample/RecyclerViewAdapter.java) as well as the [MainActivity](https://github.com/googleads/googleads-mobile-android-examples/blob/master/advanced/NativeExpressRecyclerViewExample/app/src/main/java/com/google/android/gms/example/nativeexpressrecyclerviewexample/MainActivity.java).

Also, if you haven't already, check out [this tutorial on YouTube](https://www.youtube.com/watch?v=LZCZSeFTvyk&index=11&list=PLOU2XLYxmsIKX0pUJV3uqp6N3NeHwHh0c).

Refer also to the general [Android docs for NativeExpress Ads](https://developers.google.com/admob/android/native-express).

Hope that sufficiently answers your question! Let me know if you need more clarification or I missed something.

**Comment by JonnyCiel** [comment]
Thank you for your response, but i still have problems, since i am using a different way. 

I am using a layout with de NativeExpressAdView as shown below

```
<android.support.v7.widget.CardView android:id="@+id/fragmentLostAddView"
                                    xmlns:android="http://schemas.android.com/apk/res/android"
                                    xmlns:app="http://schemas.android.com/apk/res-auto"
                                    android:layout_width="match_parent"
                                    android:layout_height="wrap_content"
                                    xmlns:ads="http://schemas.android.com/apk/res-auto"
                                    android:foreground="?android:attr/selectableItemBackground"
                                    android:minHeight="100dp"
                                    app:cardUseCompatPadding="true">


    <com.google.android.gms.ads.NativeExpressAdView
        android:layout_width="wrap_content"
        android:layout_height="wrap_content"
        android:id="@+id/nativeAd"
        android:layout_gravity="center"
        ads:adSize="350x132"
        ads:adUnitId="@string/adUnit"/>


</android.support.v7.widget.CardView>
```

Everytime the viewType is the Ad type, this layout gets inflated and i use its viewholder to call the adRequest to show the Ad. 

I tried adding the AdView to the list, but when i do that, the data into the list start adding the same views in an endless loop.

Is there a way i can display all the information in my dataset w/o being skipped when the Adview viewType is called?

Thank you, sir.

**Comment by stowy** [comment]
Hi, no problems. 

I would review the linked tutorial and sample in more detail, as you will note that the layout does not contain the adView itself. But perhaps this is not the core issue.

You stated:

> I tried adding the AdView to the list, but when i do that, the data into the list start adding the same views in an endless loop.

I would suggest that the code for that is the core of your problem. If you are getting an endless loop, have a  look at how your loop is constructed and try to understand where the loop is getting stuck. 

If you don't want to store the adViews in the list, you need to at least store some placeholder in the list that represents the fact that that item will be an ad, can just be a string for example, otherwise the item where the ad is displayed will mean that the data at that location is skipped, as you are reporting.  You need consistency between how often your adapter expects an item with viewType ad, and the items in your list that the adapter is displaying. If there is not an ad in the list where the adapter expects one, you will lose the data at that location. 

I would recommend that you post your code to the [developer technical discussion forum](https://groups.google.com/forum/#!categories/google-admob-ads-sdk/android) where you can get detailed feedback on your code. This issues location is primarily reserved for issues with the sample applications hosted here, you will find you are able to get quicker and more detailed feedback at the above forum (located at https://groups.google.com/forum/#!categories/google-admob-ads-sdk/android).  You can include a link to this issue to provide some context, but I suggest you include your code where you are populating your list for the adapter.

**Comment by JonnyCiel** [comment]
I already fix this issue.

I keep my way instead of following the youtube video tutorial, however i have to thank you, since the idea to fix my issue, was because you answer me very clear.

Thaaaaaaaank yoou

**Comment by stowy** [comment]
Great! No problems, glad you fixed it. 👍
